contracts/FeeManager.vy

```diff
@@ -0,0 +1,144 @@
+    assert not Compass(_compass).slc_switch(), "SLC is unavailable"
+
+@external
+@payable
+def deposit(depositor_paloma_address: bytes32):
+    # Deposit some balance on the contract to be used when sending messages from Paloma.
+    # depositor_paloma_address: paloma address to which to attribute the sent amount
+    self.compass_check(self.compass)
+    self.funds[depositor_paloma_address] = unsafe_add(self.funds[depositor_paloma_address], msg.value)
+    self.total_funds = unsafe_add(self.total_funds, msg.value)
+
+@internal
+def swap_grain(_grain: address, amount:uint256, dex: address, payload: Bytes[1028], min_grain: uint256) -> uint256:
+    assert min_grain > 0, "Min grain must be greater than 0"
+    grain_balance: uint256 = ERC20(_grain).balanceOf(self)
+    raw_call(dex, payload, value=amount)
+    grain_balance = ERC20(_grain).balanceOf(self) - grain_balance
+    assert grain_balance >= min_grain, "Insufficient grain received"
+    return grain_balance
+
+@external
+@nonreentrant('lock')
+def withdraw(receiver: address, amount:uint256, dex: address, payload: Bytes[1028], min_grain: uint256):
+    # Withdraw ramped up claimable rewards from compass. Withdrawals will be swapped and
+    # reimbursed in GRAIN.
+    # receiver: the validator address to receive grain token
+    # amount: the amount of COIN to withdraw.
+    # dex: address of the DEX to use for exchanging the token
+    # payload: the function payload to exchange ETH to grain for the dex
+    # min_grain: expected grain amount getting from dex to prevent front-running(high slippage / sandwich attack)
+    _compass: address = self.compass
+    self.compass_check(_compass)
+    self.claimable_rewards[receiver] = unsafe_sub(self.claimable_rewards[receiver], amount)
+    self.total_claims = self.total_claims - amount
+    assert self.claimable_rewards[receiver] >= amount, "Missing claimable rewards"
+    _grain: address = self.grain
+    grain_balance: uint256 = self.swap_grain(_grain, amount, dex, payload, min_grain)
+    ERC20(_grain).transfer(receiver, grain_balance)
+
+@external
+@payable
+def security_fee_topup():
+    _compass: address = self.compass
+    self.compass_check(_compass)
+    # Top up the security wallet with the given amount.
+    self.rewards_security_balance = unsafe_add(self.rewards_security_balance, msg.value)
+
+@external
+def transfer_fees(fee_args: FeeArgs, relayer: address):
+    # Transfer fees to the community and security wallets.
+    # fee_args: the FeeArgs struct containing the fee amounts.
+    # relayer_fee: fee to message relayer
+    # relayer: relayer address
+    _compass: address = self.compass
+    self.compass_check(_compass)
+    self.rewards_community_balance = unsafe_add(self.rewards_community_balance, fee_args.community_fee)
+    self.rewards_security_balance = unsafe_add(self.rewards_security_balance, fee_args.security_fee)
+    self.claimable_rewards[relayer] = unsafe_add(self.claimable_rewards[relayer], fee_args.relayer_fee)
```

Comment:
These values will still have to be multiplied with the current gas price.

Reply:
I thought all the fee_args are in gas cost, not gas amount. community fee in gas amount doesn't make sense. How do you determine them? @byte-bandit

Reply:
We are going to run `estimate_gas` for every TX we send. If I understand correctly, this should give us the `gas_amount` we would need for the transaction. To make sure we simulate the transaction going through all the way, we will use dummy values for the `gas_estimate` and `fee` fields. 

This is the value we will put in `gas_estimate`. It will be slightly higher as we slap a security margin on top, but that can be addressed later when we realize it works.

For the fees, we are multiplying this `gas_estimate` value with the fee settings stored on Paloma: 

- So, relayer fee could be `gas_estimate * 1.15`. 
- Security and community fees are fractions of this relayer fee, so their value could be something like `relayer_fee * 0.3` or `relayer_fee * 0.01`

So ultimately, all of these values are derived from the response we get from the `estimate_gas` RPC call. So it's all gas amount, not gas cost. Which means all of these values would need to be multiplied with the current gas cost of the TX to get the total amount of COIN needed, right?

Reply:
Yes, understood.